espnet/nets/pytorch_backend/e2e_asr_aligndenoise.py

```diff
@@ -0,0 +1,243 @@
+        group = parser.add_argument_group("aligndenoise specific setting")
+
+        group.add_argument(
+            "--aligndenoise-beta",
+            default=0.3,
+            type=float,
+        )
+        group = add_arguments_conformer_common(group)
+
+        return parser
+
+    def __init__(self, idim, odim, args, ignore_id=-1):
+        """Construct an E2E object.
+
+        :param int idim: dimension of inputs
+        :param int odim: dimension of outputs
+        :param Namespace args: argument Namespace containing options
+        """
+        super().__init__(idim, odim, args, ignore_id)
+        assert 0.0 <= self.mtlalpha < 1.0, "mtlalpha should be [0.0, 1.0)"
+
+        self.reset_parameters(args)
+
+        self.betas = np.linspace(0.08, 0.2, 50)
+        self.alphas = 1.0 - self.betas
+        self.alphas_bar = torch.tensor(np.concatenate(([1], np.cumprod(self.alphas))), dtype=torch.float32)
+
+        self.att_ctc = CTC(odim, args.adim, args.dropout_rate, ctc_type=args.ctc_type, reduce=True, include_fc=False)
+
+    def diffusion4(self, x, var, mask, noise_level=None):
+        batch_size = x.shape[0]
+        #Uniform index
+        idx = torch.randint(0, len(self.betas), size=(batch_size,))
+        lb = self.alphas_bar[idx + 1]
+        ub = self.alphas_bar[idx]
+        if not noise_level:
+            noise_level = torch.rand(size=(batch_size,)) * (ub - lb) + lb
+            noise_level = noise_level.unsqueeze(-1).unsqueeze(-1).to(device=x.device)
+        noisy_x = torch.sqrt(noise_level) * x + torch.sqrt(1.0 - noise_level) * torch.randn_like(x) * var
+        return x * mask + noisy_x * (1 - mask)
+    
+    def forward(self, xs_pad, ilens, ys_pad):
+        """E2E forward.
+
+        :param torch.Tensor xs_pad: batch of padded source sequences (B, Tmax, idim)
+        :param torch.Tensor ilens: batch of lengths of source sequences (B)
+        :param torch.Tensor ys_pad: batch of padded target sequences (B, Lmax)
+        :return: ctc loss value
+        :rtype: torch.Tensor
+        :return: attention loss value
+        :rtype: torch.Tensor
+        :return: accuracy in attention decoder
+        :rtype: float
+        """
+        # 1. forward encoder
+        xs_pad = xs_pad[:, :max(ilens)]  # for data parallel
+        src_mask = make_non_pad_mask(ilens.tolist()).to(xs_pad.device).unsqueeze(-2)
+        hs_pad, hs_mask, _ = self.encoder(xs_pad, src_mask)
+        
+        # TODO(karita) show predicted text
+        # TODO(karita) calculate these stats
+        cer_ctc = None
+        
+        batch_size = xs_pad.size(0)
+        hs_len = hs_mask.view(batch_size, -1).sum(1)
+        loss_ctc = self.ctc(hs_pad.view(batch_size, -1, self.adim), hs_len, ys_pad)
+        ys_hat = self.ctc.argmax(hs_pad.view(batch_size, -1, self.adim)).data
+        cer_ctc = self.error_calculator(ys_hat.cpu(), ys_pad.cpu(), is_ctc=True)
+
+        
+        if not self.training:
+            self.ctc.softmax(hs_pad)
```

Comment:
No tensor is assigned to the return value.

Reply:
https://github.com/espnet/espnet/blob/master/espnet/nets/pytorch_backend/e2e_asr_transformer.py#L226

It is for visualization. I don't understand the design either lol.